base/setup/lib/fsutil.c

```diff
@@ -1025,125 +1059,44 @@ DoChecking(
     return Result;
 }
 
-static BOOLEAN
-GetNextUnformattedPartition(
-    IN PPARTLIST List,
-    OUT PPARTENTRY *pPartEntry)
+static
+PVOLENTRY
+GetNextUnformattedVolume(
+    _In_ PPARTLIST List,
+    _In_opt_ PVOLENTRY Volume)
 {
-    PLIST_ENTRY Entry1, Entry2;
-    PDISKENTRY DiskEntry;
-    PPARTENTRY PartEntry;
+    PLIST_ENTRY Entry;
 
-    for (Entry1 = List->DiskListHead.Flink;
-         Entry1 != &List->DiskListHead;
-         Entry1 = Entry1->Flink)
+    for (;;)
     {
-        DiskEntry = CONTAINING_RECORD(Entry1,
-                                      DISKENTRY,
-                                      ListEntry);
+        /* If we have a current volume, get the next one, otherwise get the first */
+        Entry = (Volume ? &Volume->ListEntry : &List->VolumesList);
```

Comment:
This is actually a flattened loop that you're seeing here, so that this new version of `GetNextUnformattedVolume` can be called by using a "restarting" pointer (the `Volume` pointer, in entry), for example:

``` c
PVOLENTRY Volume = NULL;
while ((Volume = GetNextUnformattedVolume(PartitionList, Volume))
{
    /* Do something with Volume */
}
```
without having to revisit all the previous volumes in the list (for which formatting may have been skipped), which is an inconvenience the previous function suffered from.